### PR TITLE
common: WriteFileAtomic should not ignore Write error; make named return

### PR DESCRIPTION
**Pull request by odeke-em**
Fixes https://github.com/tendermint/tmlibs/issues/159

On encountering an error after file.Write, instead of returning
all along we've been proceeding with the code below it, invoking
* file.Close()
* os.Chmod
while ignoring their errors ironically, because *Atomic implies
"all or nothing" but that's contrary to the current behavior.

This change fixes that with a named return and returning early.

### CODE REVIEW

File: common/os.go [hunk 1]
```diff
@@ -126,30 +126,42 @@ func MustWriteFile(filePath string, contents []byte, mode os.FileMode) {
...
+
+	defer func() {
+		// Any error should result in full cleanup.
+		// See Issue https://github.com/tendermint/tmlibs/issues/159
+		if err != nil {
+			// If the file happens to be open, let's close it
+			// lest be hung up on it such as on Windows
+			// where we need to first close before removing
+			// See Issue https://github.com/golang/go/issues/16296.
+			if f != nil {
```
Comment (melekes) [inline common/os.go]: nice find! why just don't add this to 

```
// any err should result in full cleanup
if err != nil {
  <HERE>
  os.Remove(f.Name())	
}
```

**Comment by melekes** [comment]
What's the benefit of returning early? a few nanoseconds? plus I bet the old version is actually faster because now we have a `defer`.

**Comment by melekes** [comment]
Plus if we're gonna change this code which is battle tested, we should add more tests `62.48% (-0.21%) compared to 066fe82`

**Comment by odeke-em** [comment]
>What's the benefit of returning early? a few nanoseconds? plus I bet the old version is actually faster because now we have a defer.

Not speed related at all, it is so that we can "ATOMICALLY"(all or nothing) check for any failures e.g if f.Write fails, we should exit. If f.Close() fails we should exit, if the permission change fails, return early.

**Comment by odeke-em** [comment]
Ping @melekes @ebuchman.

**Comment by melekes** [comment]
> "ATOMICALLY"(all or nothing)

this is true for the current version

File: common/os.go (hunk 1 above)
Comment (odeke-em) [reply common/os.go]: The goal here is to close if the file is non-nil before os.Remove. I should have instead moved `os.Remove` in the `if f != nil` check

**Comment by xla** [comment]
Can we close this in favour of #171 ?

**Comment by ebuchman** [comment]
Thanks @odeke-em - let's go with #171

**Pull request closed without merging**